### DISCUSSION TITLE
Rendering weird (offline map)

### DISCUSSION THREAD

**Discussion by PhungVanHoa** [mapsforge/vtm#955] [General]
it happen on both Android and Desktop

- inconsistent: [Desktop](https://www.dropbox.com/s/kr6q4k2fw7qd32z/desktop_inconsistent.mp4?dl=0), [Android](https://www.dropbox.com/s/m2j9xs9yiaakeol/android_inconsistent.mp4?dl=0)
- run around: [Desktop](https://www.dropbox.com/s/0uij8hl1vczyup9/desktop_run_around.mp4?dl=0), [Android](https://www.dropbox.com/s/jormlkdlllkz08m/android_run_around.mp4?dl=0)

Offline Map: v5/asia/vietnam.map
Coordinates: 16.475115, 107.575097

**Comment by devemux86** [comment]
_GitHub can host images, so please include them in the report, instead of external sources._

**Reply by PhungVanHoa** [reply]
Sorry, i post video for demonstration

**Comment by devemux86** [comment]
It is strange, we've been using VTM with Mapsforge maps for years on Android and Desktop and never seen such issues.

In reports please always post the device specifications, operating system, graphics card, drivers, etc.

**Comment by devemux86** [comment]
Also report with small offline maps, like Andorra or Berlin.
(with default themes)

Then it will be easier for other users to check the reports.

**Comment by devemux86** [comment]
> Sorry, i post video for demonstration

_GitHub can also host videos._

**Comment by PhungVanHoa** [comment]
I tested on samsung note 20 ultra android 11 (pure android) and Macbook pro radeon pro 555x. I will try with map which you suggest. But can you test on area which has some Attraction. This issues only happen on that

**Comment by devemux86** [comment]
Symbols on polygons are automatically placed, enabled with `Parameters.POLY_SYMBOL`.

- Maybe the centroid algorithm in `LabelTileLoaderHook` needs improvement
- Or maybe there is a side effect on polygon symbols from Tile / Line clippers

Alternatively (or better):
- Enable the `Parameters.POLY_LABEL` for optimal placement of labels / symbols on polygons
(more expensive operation)

**Comment by PhungVanHoa** [comment]
should we precompute Poly_Label and Poly_Symbol then add them as normal point label when build map ?

**Comment by devemux86** [comment]
The flags in `Parameters` are for the VTM map library / reader.

For offline maps creation, you can see the [Mapsforge map-writer documentation](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md).

There is the `Parameters.POLY_CENTROID` for centroid pre-calculation when reading the Mapsforge maps.
(more expensive operation)

For more details, it is easier to check where each parameter is used inside the code to understand its use.
(for most users the default values are sufficient)

**Reply by PhungVanHoa** [reply]
Thank you

**Comment by PhungVanHoa** [comment]
i disable POLY_SYMBOL (POLY_LABEL is false default) but symbol still display inconsistent. Enable POLY_CENTROID not working


https://user-images.githubusercontent.com/16158238/185155792-de182fcb-89d4-415a-8f38-549eb3f2658a.mp4

**Comment by devemux86** [comment]
`Parameters.POLY_SYMBOL` enables the auto-generated symbols on polygons.
The symbols on nodes are always present.

Try with the `Parameters.POLY_LABEL` (to change the centroid calculation).
One (attraction) symbol near Berlin label was "fixed" with that parameter.

(all tests with Mapsforge offline maps)

**Reply by PhungVanHoa** [reply]
I did, but result is the same :(. why attraction symbol still display when i disable Parameters.POLY_SYMBOL

**Comment by devemux86** [comment]
> result is the same

Someone with free time can check the symbols' origin (nodes | polygons),
why the "attraction" behave like that, what happens on other platforms, etc.

> why attraction symbol still display when i disable Parameters.POLY_SYMBOL

It is already explained [above](https://github.com/mapsforge/vtm/discussions/955#discussioncomment-3415380): symbols exist on nodes and auto-generated on polygon centroids.
`Parameters.POLY_SYMBOL` disables the polygon symbols.
For more details you have to study the library and render themes.

> Parameters.POLY_SYMBOL enables the auto-generated symbols on polygons.
The symbols on nodes are always present.

Better to find an example on berlin.map so the community can check it easier.
I can test it on Desktop (Linux) or Android emulator when time permits.

**Comment by PhungVanHoa** [comment]
i did test on Berlin map (lat: 52.519597, lng: 13.360688). Arrow on road display inconsistently

https://user-images.githubusercontent.com/16158238/185177776-a57fb6f9-941b-498e-ba5e-80d758651de0.mp4

**Comment by devemux86** [comment]
> i did test on Berlin map (lat: 52.519597, lng: 13.360688). Arrow on road display inconsistently

Thanks, it can be reproduced on Desktop (Linux),
but cannot know what happens there without proper check.

**Comment by PhungVanHoa** [comment]
another issue with building in Berlin

Coordinates: 52.525918, 13.395091

https://user-images.githubusercontent.com/16158238/185186830-e0d4a569-759d-4011-81ac-61f79ae32cfb.mp4

**Comment by PhungVanHoa** [comment]
Symbol run around in offline map Berlin

https://user-images.githubusercontent.com/16158238/185282287-4acde7f8-7515-4233-83e0-b740450f8e61.mp4

**Comment by devemux86** [comment]
> Symbol run around in offline map Berlin

As mentioned [above](https://github.com/mapsforge/vtm/discussions/955#discussioncomment-3415380):

> Try with the Parameters.POLY_LABEL (to change the centroid calculation).
One (attraction) symbol near Berlin label was "fixed" with that parameter.

**Reply by PhungVanHoa** [reply]
Oh sorry. Thanks!
